src/Controller/FragmentRegistry/FragmentRegistry.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2016 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\Controller\FragmentRegistry;
+
+/**
+ * Fragment registry.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class FragmentRegistry implements FragmentRegistryInterface
+{
+    const TYPES = [
+        ContentElementInterface::class,
+        FrontendModuleInterface::class,
+        InsertTagInterface::class,
+        PageTypeInterface::class,
+    ];
+
+    /**
+     * @var array
+     */
+    private $fragments = [];
+
+    /**
+     * @var array
+     */
+    private $fragmentsPerType = [];
+
+    /**
+     * FragmentRegistry constructor.
+     */
+    public function __construct()
+    {
+        foreach (self::TYPES as $type) {
+            $this->fragmentsPerType[$type] = [];
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFragment(FragmentInterface $fragment)
+    {
+        $this->fragments[] = $fragment;
+
+        foreach (self::TYPES as $type) {
+            if (is_a($fragment, $type)) {
```

Comment:
I don't think this is a good idea, we can't add another interface in the future if we want to add new features (which we cannot add to the existing interface due to BC :-)).